File: src/python/sync_smooth.py
import time

import numpy as np
import pyautogui
import sounddevice as sd


class MouseControlledSynth:
    """マウス位置で制御されるオシレータシンク・シンセサイザー"""

    def __init__(self, samplerate=48000, time_constant_ms=100, polling_interval_ms=8):
        """
        Args:
            samplerate: サンプリングレート (Hz)
            time_constant_ms: 時定数 (ms)。この時間で目標値の約63%に到達
                - 50ms:  素早い応答
                - 100ms: 標準的な応答 (デフォルト)
                - 200ms: ゆったりした応答
            polling_interval_ms: マウス位置のポーリング間隔 (ms)
                - 8ms: 標準 (デフォルト, 125Hz)
                - 4ms: 高速 (250Hz)
                - 16ms: 低速 (62.5Hz)
        """
        self.fs = samplerate
        self.time_constant_ms = time_constant_ms
        self.polling_interval_ms = polling_interval_ms

        # 時定数から滑らかさ係数を計算
        n = time_constant_ms * samplerate / 1000
        self.smoothness_coeff = 1.0 / n

        self.blocksize = int(samplerate * polling_interval_ms / 1000)

        print("設定:")
        print(f"  時定数: {self.time_constant_ms}ms (約63%到達時間)")
        print(f"  滑らかさ係数: {self.smoothness_coeff:.6f}")
        print(f"  ポーリング間隔: {self.polling_interval_ms}ms ({1000 / self.polling_interval_ms:.1f}Hz)")
        print(f"  ブロックサイズ: {self.blocksize}サンプル ({self.blocksize / self.fs * 1000:.1f}ms)")

        # 位相管理
        self.phase_master = 0.0
        self.phase_slave = 0.0

        # 周波数管理
        self.current_freq_master = 100.0
        self.current_freq_slave = 100.0
        self.target_freq_master = 100.0
        self.target_freq_slave = 100.0

        # 画面サイズ取得
        self.screen_width, self.screen_height = pyautogui.size()

    def _callback(self, outdata, frames, time_info, status):
        """オーディオコールバック関数 : blocksize間隔で呼ばれる"""
        # 各サンプルの周波数を格納する配列
        freq_master_per_sample = np.zeros(frames, dtype=np.float32)
        freq_slave_per_sample = np.zeros(frames, dtype=np.float32)

        # 1サンプルずつ周波数を計算（指数平滑化）
        temp_freq_master = self.current_freq_master
        temp_freq_slave = self.current_freq_slave
        for i in range(frames):
            temp_freq_master += (self.target_freq_master - temp_freq_master) * self.smoothness_coeff
            temp_freq_slave += (self.target_freq_slave - temp_freq_slave) * self.smoothness_coeff
            freq_master_per_sample[i] = temp_freq_master
            freq_slave_per_sample[i] = temp_freq_slave

        # このブロックの最終的な周波数を次のブロックの開始値として保存
        self.current_freq_master = temp_freq_master
        self.current_freq_slave = temp_freq_slave

        # 位相増分を計算
        phase_inc_master = freq_master_per_sample / self.fs
        phase_inc_slave = freq_slave_per_sample / self.fs

        # 累積位相を計算
        master_phase = (self.phase_master + np.cumsum(phase_inc_master)) % 1.0
        slave_phase = (self.phase_slave + np.cumsum(phase_inc_slave)) % 1.0

        # マスター位相リセット検出 (ハードシンク)
        reset_points = np.where(np.diff(master_phase) < 0)[0] + 1

        # 各リセットポイントで、slaveの位相を0から再計算する
        for rp in reset_points:
            slave_phase[rp:] = np.cumsum(phase_inc_slave[rp:]) % 1.0

        # ノコギリ波生成 (-1.0 to 1.0)
        out = (2.0 * slave_phase - 1.0).astype(np.float32)

        # 次のブロックのために位相を更新
        self.phase_master = master_phase[-1]
        self.phase_slave = slave_phase[-1]

        outdata[:] = out.reshape(-1, 1)

    def update_target(self, x, y):
        """マウス位置から目標周波数を更新"""
        self.target_freq_master = np.interp(x, [0, self.screen_width], [40, 600])
        self.target_freq_slave = np.interp(y, [0, self.screen_height], [2000, 100])

    def run(self):
        """メインループ実行"""
        with sd.OutputStream(
            channels=1, callback=self._callback, samplerate=self.fs, blocksize=self.blocksize
        ) as stream:
            print("\nマウスを動かして音を制御してください (Ctrl+C で終了)")
            print(f"オーディオデバイスのレイテンシ: {stream.latency * 1000:.2f} ms")
            try:
                while True:
                    x, y = pyautogui.position()
                    self.update_target(x, y)
                    time.sleep(self.polling_interval_ms / 1000)
            except KeyboardInterrupt:
                print("\n終了しました。")


def main():
    synth = MouseControlledSynth(
        samplerate=48000,
        time_constant_ms=16,  # 時定数 (ms)。この時間で目標値の約63%に到達
        polling_interval_ms=8,  # マウス位置のポーリング間隔 (ms)
    )
    synth.run()


if __name__ == "__main__":
    main()

File: src/rust/src/sync_smooth.rs
// ハードシンク・オシレータ スムーズ版
// マウスの位置でマスター周波数とスレーブ周波数を制御します
// X軸: マスター周波数 (40Hz - 600Hz)
// Y軸: スレーブ周波数 (100Hz - 2000Hz)
// 指数平滑化により1サンプルごとの滑らかな周波数変化を実現
// Ctrl+C で終了してください

use cpal::traits::{DeviceTrait, HostTrait, StreamTrait};
use rdev::{display_size, listen, EventType};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

// グローバル状態
struct SynthState {
    // 目標周波数（マウス位置から計算）
    target_freq_master: f32,
    target_freq_slave: f32,
    // 現在の周波数（指数平滑化された値）
    current_freq_master: f32,
    current_freq_slave: f32,
    // 位相管理
    phase_master: f32,
    phase_slave: f32,
    // 滑らかさ係数
    smoothness_coeff: f32,
}

impl SynthState {
    fn new(time_constant_ms: f32, sample_rate: f32) -> Self {
        // 時定数から滑らかさ係数を計算
        // n = time_constant_ms * sample_rate / 1000
        // smoothness_coeff = 1.0 / n
        let n = time_constant_ms * sample_rate / 1000.0;
        let smoothness_coeff = 1.0 / n;

        println!("設定:");
        println!("  時定数: {}ms (約63%到達時間)", time_constant_ms);
        println!("  滑らかさ係数: {:.6}", smoothness_coeff);

        SynthState {
            target_freq_master: 100.0,
            target_freq_slave: 100.0,
            current_freq_master: 100.0,
            current_freq_slave: 100.0,
            phase_master: 0.0,
            phase_slave: 0.0,
            smoothness_coeff,
        }
    }
}

fn main() -> Result<(), anyhow::Error> {
    println!("🎵 Cat Oscillator Sync - Smooth Version (Rust)");
    println!("マウスを動かして音を制御してください");
    println!("X軸: マスター周波数 (40Hz - 600Hz)");
    println!("Y軸: スレーブ周波数 (100Hz - 2000Hz)");
    println!("指数平滑化により滑らかな周波数変化を実現");
    println!("Press Ctrl+C to exit");
    println!();

    // パラメータ設定
    let time_constant_ms = 16.0; // 時定数 (ms)
    let polling_interval_ms = 8; // マウス位置のポーリング間隔 (ms)

    // 画面サイズの取得
    let (screen_width, screen_height) =
        display_size().map_err(|e| anyhow::anyhow!("画面サイズの取得に失敗: {:?}", e))?;
    println!("Screen size: {}x{}", screen_width, screen_height);

    // オーディオホストとデバイスの取得
    let host = cpal::default_host();
    let device = host
        .default_output_device()
        .ok_or_else(|| anyhow::anyhow!("デフォルトの出力デバイスが見つかりません"))?;

    println!("Output device: {}", device.name()?);

    // デバイス設定の取得
    let default_config = device.default_output_config()?;
    println!("Default output config: {:?}", default_config);

    // モノラル（1チャンネル）で設定を構築
    let config = cpal::StreamConfig {
        channels: 1,
        sample_rate: default_config.sample_rate(),
        buffer_size: cpal::BufferSize::Default,
    };
    let sample_rate = config.sample_rate.0 as f32;
    println!(
        "Configured for: {} Hz, {} channel (mono)",
        sample_rate, config.channels
    );

    // blocksize の計算（ポーリング間隔に対応）
    let blocksize = (sample_rate * polling_interval_ms as f32 / 1000.0) as usize;
    println!(
        "  ポーリング間隔: {}ms ({:.1}Hz)",
        polling_interval_ms,
        1000.0 / polling_interval_ms as f32
    );
    println!(
        "  ブロックサイズ: {}サンプル ({:.1}ms)",
        blocksize,
        blocksize as f32 / sample_rate * 1000.0
    );

    // 状態の共有
    let state = Arc::new(Mutex::new(SynthState::new(time_constant_ms, sample_rate)));
    let mouse_pos = Arc::new(Mutex::new((
        screen_width as f64 / 2.0,
        screen_height as f64 / 2.0,
    )));

    // マウスイベントリスナー用の位置共有
    let mouse_pos_listener = Arc::clone(&mouse_pos);
    thread::spawn(move || {
        if let Err(error) = listen(move |event| {
            if let EventType::MouseMove { x, y } = event.event_type {
                let mut pos = mouse_pos_listener.lock().unwrap();
                *pos = (x, y);
            }
        }) {
            eprintln!("マウスリスナーエラー: {:?}", error);
        }
    });

    // マウス位置から目標周波数を計算する定期処理
    let state_updater = Arc::clone(&state);
    let mouse_pos_updater = Arc::clone(&mouse_pos);
    thread::spawn(move || {
        loop {
            let (x, y) = {
                let pos = mouse_pos_updater.lock().unwrap();
                *pos
            };

            // 周波数のマッピング
            let target_freq_master = 40.0 + (x / screen_width as f64) * (600.0 - 40.0);
            let target_freq_slave =
                100.0 + ((screen_height as f64 - y) / screen_height as f64) * (2000.0 - 100.0);

            {
                let mut s = state_updater.lock().unwrap();
                s.target_freq_master = target_freq_master as f32;
                s.target_freq_slave = target_freq_slave as f32;
            }

            thread::sleep(Duration::from_millis(polling_interval_ms));
        }
    });

    // オーディオストリーム構築
    let state_audio = Arc::clone(&state);
    let stream = device.build_output_stream(
        &config,
        move |data: &mut [f32], _: &cpal::OutputCallbackInfo| {
            let mut s = state_audio.lock().unwrap();

            // 1サンプルずつ周波数を計算（指数平滑化）
            for sample in data.iter_mut() {
                // 指数平滑化：current += (target - current) * smoothness_coeff
                s.current_freq_master +=
                    (s.target_freq_master - s.current_freq_master) * s.smoothness_coeff;
                s.current_freq_slave +=
                    (s.target_freq_slave - s.current_freq_slave) * s.smoothness_coeff;

                // 位相増分を計算
                let inc_master = s.current_freq_master / sample_rate;
                let inc_slave = s.current_freq_slave / sample_rate;

                // マスターオシレータの位相を進める
                s.phase_master += inc_master;

                // マスター位相のラップアラウンドを検出
                if s.phase_master >= 1.0 {
                    s.phase_master -= 1.0;
                    // マスターがリセットされたら、スレーブもリセット
                    s.phase_slave = 0.0;
                } else {
                    // マスターがリセットされていない場合のみ、スレーブ位相を進める
                    s.phase_slave += inc_slave;
                    if s.phase_slave >= 1.0 {
                        s.phase_slave -= 1.0;
                    }
                }

                // ノコギリ波: 位相を -1.0 から 1.0 の範囲に変換
                *sample = 2.0 * s.phase_slave - 1.0;
            }
        },
        |err| eprintln!("Stream error: {}", err),
        None,
    )?;

    // ストリーム開始
    stream.play()?;

    // 周波数表示スレッド
    let state_display = Arc::clone(&state);
    thread::spawn(move || loop {
        thread::sleep(Duration::from_millis(500));
        let s = state_display.lock().unwrap();
        print!(
            "\rMaster: {:6.1} Hz | Slave: {:6.1} Hz",
            s.current_freq_master, s.current_freq_slave
        );
        std::io::Write::flush(&mut std::io::stdout()).ok();
    });

    println!("\nPress Ctrl+C to stop...");
    std::thread::park();

    Ok(())
}
